### ISSUE TITLE
verify_active_connections! removed in Rails4

### ISSUE THREAD

**Issue by kenmazaika** [socialcast/resque-ensure-connected#3]
@fxn it seems you removed verify_active_connections! for Rails4 beta.

https://github.com/rails/rails/commit/9d1f1b1ea9e5d637984fda4f276db77ffd1dbdcb

This seems like a completely valid thing to do, as the method was marked with `:nodoc:`.

However, this patch to make resque workers reconnect to the database if they got disconnected no longer seems to apply.

Would switching this line to something like:

```
ActiveRecord::Base.connection_pool.connections.each {|c| c.verify! }
```

work in Rails4?  Also could it cause a problem that it isn't `synchronized`?

**Comment by codyrobbins** [comment]
Yeah, this is problematic because the [Resque FAQ](https://github.com/resque/resque/wiki/FAQ#how-do-you-work-around-the-mysql-server-has-gone-away-error-) still indicates that `verify_active_connections!` is the way to work around the dropped connections problem. The commit message by @fxn simply says that `verify_active_connections!` is obsolete and that Rails does it a “different way” now, but there’s no mention of what that way is.

@fxn: Could you comment on what `verify_active_connections!` should be replaced with?

**Comment by fxn** [comment]
I believe what I had in mind in the commit message was that Rails handles this differently nowadays, there is a middleware that returns connections to the pool, instead of the old dispatcher performing that call.

From the source code I'd say that connections are monitored to be active, but let me summon @tenderlove, he has worked in this part.

**Comment by codyrobbins** [comment]
Thanks very much for the quick response, @fxn. Much appreciated!

**Comment by ghiculescu** [comment]
Hi - just wondering if there was any follow up here?

**Comment by fxn** [comment]
Nowadays there is a middleware called `ActiveRecord::ConnectionApadaters::ConnectionManagement` that runs `ActiveRecord::Base.clear_active_connections!` after its own invocation of `call` returns. That call checks in all connections in the calling thread to their respective connection pools. Normally there is only one connection pool and in that case we are talking about checking in the unique connection in the current thread, if any.

When a connection is needed by the current thread and it has not been yet checked out, it is transparently checked out for you from the pool and verified. Is in that sense that I said this verification is currently built-in in Rails, instead of someone external to AR like the old dispatcher verifying by hand.

Therefore, I believe the FAQ could suggest something like this:

``` ruby
Resque.after_fork do
  ActiveRecord::Base.clear_active_connections!
end
```

That plays well with the way connection handling is done nowadays.

**Comment by rebyn** [comment]
I wonder if [`ActiveRecord::Base.connection.active?`](http://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/AbstractAdapter.html#method-i-active-3F) now does the job?

**Update**: it seems so:

```
# after shutting down MySQL
> ActiveRecord::Base.connection.active?
 => false
```

**Comment by ioquatix** [comment]
Any update on this?

**Comment by memosstilvi** [comment]
bump?

**Comment by zimini** [comment]
we could manage the connection through the _ActiveRecord::Base.clear_active_connections!_ It seems it makes the same job as _ActiveRecord::Base.verify_active_connections!_ (Rails 3)

**Comment by chrishough** [comment]
bump! cc: @ideaoforder @dza2000
